New: Date Picker Component

davepoon opened the pull request.

omgaz commented on src/components/Main.js:
```diff
@@ -53,9 +55,14 @@ class AppComponent extends React.Component {
   }
 
+  setSelectedDate(newValue) {
+    this.setState({ startDate: newValue });
```
Need coverage on this line.

omgaz commented on src/components/Main.js:
```diff
@@ -419,6 +426,12 @@ class AppComponent extends React.Component {
 
+        <h1>Date Picker</h1>
+        <DatePicker
+          dateFormat="DD/MM/YYYY"
```
Preferred format e.g. '14 Apr 2016'

omgaz commented on src/styles/_react-datepicker-custom.scss:
```diff
@@ -0,0 +1,70 @@
+
+  &__input-container {
+
+    input {
```
Most of this, and the variables, can be resolved if we can add a class to the input `.form-control`

davepoon commented:
All the issues have been addressed.
- Reuse bootstrap .form-control styling.
-  Date format is changed to '14 Apr 2016'.
- The new design is added to the datepicker.
- The design has been reviewed by Pedros.
- Code coverage is now 100%.

tchcxp commented on test/components/MainTest.js:
```diff
@@ -138,4 +139,15 @@ describe('MainComponent', () => {
+  it('should toggle `onChange` on `Open DatePicker` change', () => {
+    MainComponent.setState({ startDate: moment() });
+    const datePickerComponent = MainComponent.find(DatePicker);
+    datePickerComponent.find('.form-control').simulate('change');
```
We need to check the change after clicking the component (eg. change of date). Otherwise, it adds code coverage but does not test the behaviour.

davepoon replied:
@tchcxp this is not a button, it's an input field, so I had to use change instead of click to simulate the interaction. The existing components are clicking a button to trigger the action.

davepoon commented:
This is the screenshot of the design.
<img width="413" alt="datepicker2" src="https://cloud.githubusercontent.com/assets/138992/15131822/a9159968-1696-11e6-9516-607d103543b7.png">

tchcxp commented:
LGTM when you can add the small test for calling date picker component as we discussed. :)

davepoon commented:
The small test is added.
@omgaz @tchcxp welcome for code review.

omgaz commented on src/components/Main.js:
```diff
@@ -53,9 +55,14 @@ class AppComponent extends React.Component {
 
     this.state = {
       showSimpleModal: false,
+      startDate: moment(),
```
I'd rather abstract out the implementation of moment so it's a worker rather than constructor.

`startDate: moment().toDate(),` would be preferred.

@etaoins ? thoughts?

etaoins replied:
Agreed. Lets keep the model using JavaScript dates.

davepoon replied:
It's simpler, but React Date Picker component doesn't parse javascript date object properly, it largely depends on moment object to function properly.

omgaz replied:
Cool, yea that's horrible https://github.com/Hacker0x01/react-datepicker/blob/master/src/day.jsx#L41

`// React datepicker expects a moment date, rather than JS date`

omgaz replied:
^^ can we get a comment then I'm happy to merge :) WOOHOO!

davepoon replied:
I almost missed this hidden comment...

omgaz commented on src/styles/_react-datepicker-custom.scss:
```diff
@@ -0,0 +1,146 @@
+$datepicker-border-radius: $border-radius-base;
+$datepicker-header-color-background: $color-white;
+$datepicker-selected-colour-background: $color-primary;
+$datepicker-color-disabled: $color-gray;
```
Do we not have a $color-disabled variable? If not we should probably add one to variable/color scss file and use that instead of the colour directly

omgaz replied:
Move all the variables inside the scope of .react-datepicker if we can (down two lines)

omgaz commented on src/styles/_react-datepicker-custom.scss:
```diff
@@ -0,0 +1,146 @@
+$datepicker-font-weight: $font-weight-medium;
+$datepicker-font-size: $font-size-base;
+$datepicker-border-radius: $border-radius-base;
+$datepicker-header-color-background: $color-white;
```
Do we have a $color-background we can use instead of white directly

davepoon replied:
The calendar background is white colour.

omgaz commented on src/styles/_react-datepicker-custom.scss:
```diff
@@ -0,0 +1,146 @@
+$datepicker-color-disabled: $color-gray;
+
+.react-datepicker {
+  box-shadow: 0 2px 0 $color-gray;
```
Ditto on the $color-gray, we shouldn't be using these directly if we can help it. Maybe create a color-shadow variable

omgaz commented:
`src/styles/color.scss@35: // Usage: Use these values, not the colours directly`

omgaz commented on src/styles/_react-datepicker-custom.scss:
```diff
@@ -0,0 +1,146 @@
+
+.react-datepicker {
+  box-shadow: 0 2px 0 $color-gray;
+  color: $color-gray-darker;
```
This is a form element: `$color-text-form`

omgaz commented on src/styles/_react-datepicker-custom.scss:
```diff
@@ -0,0 +1,146 @@
+  }
+
+  &__day {
+    color: $color-gray-darker;
```
remove these two in place of

&
&__day
&__day-name {
  color: $color-text-form;
}

omgaz commented on src/styles/_react-datepicker-custom.scss:
```diff
@@ -0,0 +1,146 @@
+    }
+
+    &__day--in-range {
+      color: $color-white;
```
$color-inverse

omgaz commented on src/styles/_react-datepicker-custom.scss:
```diff
@@ -0,0 +1,146 @@
+
+    &--selected {
+      background-color: $datepicker-selected-colour-background;
+      color: $color-white;
```
$color-inverse

omgaz commented on src/styles/_react-datepicker-custom.scss:
```diff
@@ -0,0 +1,146 @@
+      content: '';
+      display: block;
+      height: 24px;
+      width: 24px;
```
24px is used a lot, make this a variable please

omgaz commented on src/styles/_react-datepicker-custom.scss:
```diff
@@ -0,0 +1,146 @@
+    right: 7px;
+
+    &:hover {
+      background-color: $color-gray-lighter;
```
Again, $color-background is lightest, can we not keep it consistent.

omgaz commented on src/styles/_react-datepicker-custom.scss:
```diff
@@ -0,0 +1,146 @@
+  }
+
+  &__current-month {
+    color: $color-gray-darker;
```
$color-text-form

davepoon commented:
@omgaz It's been updated regarding to your feedback and addressed the issues you mentioned.

davepoon commented:
Anymore feedback? :)

davepoon commented:
@omgaz @tchcxp Feedback? If no, I gonna rebase it.

davepoon commented:
Fixed background-position issue because Firefox doesn't support background-position-x
Fixed the '"clear button" in Internet Explorer 10 and 11 and Edge 12+.

etaoins commented:
LGTM

davepoon commented:
Merging?

omgaz commented on src/styles/_react-datepicker-custom.scss:
```diff
@@ -0,0 +1,147 @@
+    &::before {
+      border: 0;
+    }
+  }
```
This can be re-written as:

```scss
&__tether-element-attached-top &__triangle {
  &,
  &::before {
    border: 0;
  }
}
```

davepoon replied:
Good catch, the ampersand is such a great feature in sass, thanks :)

omgaz commented on src/styles/color.scss:
```diff
@@ -85,3 +85,6 @@ $color-border: $color-gray;
 $color-shadow: $color-black;
+
+// Modal
+$color-modal-shadow: rgba($color-shadow, .25);
```
👍

omgaz commented on src/styles/_react-datepicker-custom.scss:
```diff
@@ -0,0 +1,147 @@
+  $datepicker-font-size: $font-size-base;
+  $datepicker-icon-size: 24px;
+  $datepicker-border-radius: $border-radius-base;
+  $datepicker-header-color-background: $color-inverse;
```
I don't think `inverse` is correct. The datepicker isn't an inverted colour scheme. I'd use `$popover-bg`

omgaz commented on src/styles/_react-datepicker-custom.scss:
```diff
@@ -0,0 +1,147 @@
+    right: 7px;
+
+    &:hover {
+      background-color: $color-disabled;
```
Hover states aren't disabled even if they co-incidentally share the same colour. If the context is not correct don't use it, create your own colour context variable at the top of the file and map directly to the colour that represents.

$datepicker-selected-bg-color: $color-gray-lighter

...

background-color: $datepicker-selected-bg-color;

omgaz commented on src/styles/_react-datepicker-custom.scss:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    &--selected {
+      background-color: $datepicker-selected-colour-background;
```
Awesome. So you get that this is a datepicker selected state so you have a context that makes sense.

omgaz commented on src/styles/_react-datepicker-custom.scss:
```diff
@@ -0,0 +1,147 @@
+    left: 7px;
+
+    &:hover {
+      background-color: $color-disabled;
```
ditto

omgaz commented on src/styles/_react-datepicker-custom.scss:
```diff
@@ -0,0 +1,147 @@
+    border: 0;
+
+    &::before {
+      border: 0;
```
ditto, DRY (don't repeat yourself) :)

omgaz commented:
It's ok to hate me by this point :)

You're doing great though

davepoon commented:
No, I like all your comments, which make a lot of sense. This is a great way to learn and improve code quality.

davepoon commented:
@omgaz All have been updated.

omgaz commented:
Do you just wanna add the line comment on `src/components/Main.js` about moment and then I'll merge. ✋ high five gg

davepoon commented:
@omgaz NO!!......problem. Done! Merge me! Merge me!!

omgaz commented:
Merging :ramen:

omgaz commented:
Awesome!

davepoon commented:
OMG! It's merged! I can't get to sleep tonight!! :flushed: Too excited!! :tada:
